engines/query-sparql/test/QuerySparql-test.ts

```diff
@@ -334,6 +334,90 @@ describe('System test: QuerySparql', () => {
           .toEqual(1);
       });
 
+      it('with insert where on a single source', async() => {
+        // Prepare store
+        const store = new Store([
+          DF.quad(DF.namedNode('ex:s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()),
+          DF.quad(DF.blankNode('s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()),
+          DF.quad(DF.blankNode('ex:s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()),
+        ]);
+
+        // Execute query
+        const result = <RDF.QueryVoid> await engine.query(`INSERT {
+          ?s <ex:a> <ex:thing> .
+          ?p <ex:a> <ex:thing> .
+          ?o <ex:a> <ex:thing> .
+        } WHERE { ?s ?p ?o }`, {
+          sources: [ store ],
+        });
+        await result.execute();
+
+        // Check store contents
+        expect(store.size).toEqual(8);
+        expect(store.countQuads(DF.namedNode('ex:s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()))
+          .toEqual(1);
+        expect(store.countQuads(DF.blankNode('s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()))
+          .toEqual(1);
+        expect(store.countQuads(DF.blankNode('s'), DF.namedNode('ex:a'), DF.namedNode('ex:thing'), DF.defaultGraph()))
+          .toEqual(1);
+        expect(
+          store.countQuads(DF.blankNode('ex:s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()),
+        )
+          .toEqual(1);
+        expect(
+          store.countQuads(DF.blankNode('ex:s'), DF.namedNode('ex:a'), DF.namedNode('ex:thing'), DF.defaultGraph()),
+        )
+          .toEqual(1);
+        expect(
+          store.countQuads(DF.namedNode('ex:s'), DF.namedNode('ex:a'), DF.namedNode('ex:thing'), DF.defaultGraph()),
+        )
+          .toEqual(1);
+        expect(
+          store.countQuads(DF.namedNode('ex:p'), DF.namedNode('ex:a'), DF.namedNode('ex:thing'), DF.defaultGraph()),
+        )
+          .toEqual(1);
+        expect(
+          store.countQuads(DF.namedNode('ex:o'), DF.namedNode('ex:a'), DF.namedNode('ex:thing'), DF.defaultGraph()),
+        )
+          .toEqual(1);
+      });
+
+      it('with insert where on two sources', async() => {
+        // Prepare store
+        const store = new Store([
+          DF.quad(DF.blankNode('ex:s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()),
+        ]);
+        const store2 = new Store([
+          DF.quad(DF.blankNode('ex:s'), DF.namedNode('ex:p'), DF.namedNode('ex:o'), DF.defaultGraph()),
+        ]);
+
+        // Execute query
+        const result = <RDF.QueryVoid> await engine.query(`INSERT {
+          ?s <ex:a> <ex:thing> .
+        } WHERE { ?s ?p ?o }`, {
+          sources: [ store, store2 ],
+          destination: store,
+          [KeysRdfResolveQuadPattern.sourceIds.name]: new Map(),
+        });
+        await result.execute();
+
+        // Check store contents
+        expect(store.size).toEqual(3);
```

Comment:
Could you add a comment line here to explain the expected results?
Because it took me a while to understand what is going on 😅 

So if I understand correctly, two identical quads from separate sources produce 2 results. One quad originates from the destination, so that blank node becomes the original one. But the other quad is from another source, so that one receives a new bnode label (the skolemized one).

Reply:
Yep - your understanding is correct, I'll add a comment :)